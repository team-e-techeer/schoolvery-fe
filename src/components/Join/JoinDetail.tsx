import { css } from '@emotion/react';
import { BiCategory as CategoryIcon } from 'react-icons/bi';
import { IoPeopleCircle as People, IoTimeOutline as Time } from 'react-icons/io5';
import { FaMoneyCheck as CardIcon } from 'react-icons/fa';
import colors from '@/constants/colors';
import Button from '../Button';
import { useRecoilValue } from 'recoil';
import { watchingPostState } from '@/atoms/user/watchingPostState';
import { useTimer } from '@/hooks/useTimer';
import dayjs from 'dayjs';
import 'dayjs/locale/ko';
import { useGetChatInfoWithPostId } from '@/hooks/query/chat/GET/useGetChatInfoWithPostId';
import { userState } from '@/atoms/user/userState';
import { useCallback, useEffect } from 'react';
import { usePostParticipateChat } from '@/hooks/query/chat/POST/usePostParticipateChat';
import { useNavigate } from 'react-router-dom';
dayjs.locale('ko');

const spanStyle = css`
  display: flex;
  align-items: center;
  font-size: 1.3rem;
  margin-top: 1rem;
`;

const ulStyle = css`
  display: flex;
  align-items: center;
  font-size: 1.3rem;
  margin-top: 1rem;
`;

const eachSpanStyle = css`
  display: flex;
  flex: 1;
  font-size: 1.3rem;
  :nth-of-type(2) {
    color: ${colors.blue500};
  }
`;

const iconStyle = css`
  margin-right: 2rem;
  font-size: 2rem;
  color: ${colors.mainColor};
`;

interface Props {
  shopName: string;
  schoolName: string;
  categoryName: string;
  people: {
    current: number;
    total: number;
  };
  time: {
    post: number;
    left: number;
  };
  payment: number;
}

export default function JoinDetail({ shopName, schoolName, categoryName, people, time, payment }: Props) {
  const { accessToken, id } = useRecoilValue(userState);
  const watchingPost = useRecoilValue(watchingPostState);
  const chatPageInfo = useGetChatInfoWithPostId({ accessToken, postId: watchingPost.id });

  const enterChatMutate = usePostParticipateChat();
  const onClickEnter = useCallback(() => {
    chatPageInfo.data &&
      enterChatMutate.mutate({
        accessToken,
        postData: { room_id: chatPageInfo.data?.id, user_id: id, name: watchingPost.title },
      });
  }, [chatPageInfo.data, userState]);
  const navigate = useNavigate();

  useEffect(() => {
    if (enterChatMutate.isSuccess && enterChatMutate.data?.room_id) {
      navigate(`/chat/?room=${enterChatMutate.data?.room_id}`, {
        state: { roomId: enterChatMutate.data.room_id, storeName: watchingPost.title },
        replace: true,
      });
    }
  }, [enterChatMutate.isSuccess, enterChatMutate.data?.room_id]);

  return (
    <>
      <div
        onClick={onClickEnter}
        css={css`
          display: flex;
          flex-direction: column;
          width: 90%;
          margin: 0 auto;
          background-color: ${colors.white};
          border-radius: 8px;
          padding: 2rem;
          margin-top: 10rem;
          box-shadow: 2px 2px 2px 2px ${colors.grey300};
          -webkit-font-smoothing: antialiased;
          transition: color 0.1s ease-in-out, background-color 0.1s ease-in-out;
          &:active {
            background-color: ${colors.grey300};
            box-shadow: 2px 2px 2px 2px ${colors.grey200};
          }
        `}
      >
        <h1
          css={css`
            display: inline-block;
            width: max-content;
            font-size: 2rem;
            margin-bottom: 1rem;
            background-color: ${colors.subColor};
            padding: 1rem;
            border-radius: 0.7rem;
            color: ${colors.white};
          `}
        >
          {watchingPost.store}
        </h1>

        <span css={spanStyle}>
          <CategoryIcon css={iconStyle} />
          {categoryName}
        </span>
        <ul css={ulStyle}>
          <People css={iconStyle} />
          <span css={eachSpanStyle}>{watchingPost.peopleNum}명</span>
          <span css={eachSpanStyle}>현재 {people.current}명</span>
        </ul>
        <ul css={ulStyle}>
          <Time css={iconStyle} />
          <span css={eachSpanStyle}>{dayjs(watchingPost.deadline).format('A hh : mm')}</span>
          <span css={eachSpanStyle}>{watchingPost.left} 남음</span>
        </ul>
        <ul css={ulStyle}>
          <CardIcon css={iconStyle} />
          <span css={eachSpanStyle}>총 {watchingPost.deliveryFee}원</span>
          <span css={eachSpanStyle}>인당 {Math.round(watchingPost.deliveryFee / watchingPost.peopleNum)}원 예상</span>
        </ul>
        {watchingPost.content && (
          <div
            css={css`
              width: 100%;
              background-color: #eae9f8;
              margin-top: 2rem;
              border-radius: 0.7rem;
              padding: 2rem;
            `}
          >
            <p
              css={css`
                font-size: 1.3rem;
                color: ${colors.black};
              `}
            >
              {watchingPost.content}
            </p>
          </div>
        )}

        <Button
          onClick={onClickEnter}
          css={css`
            display: flex;
            align-self: flex-end;
            margin: 0;
            margin-top: 2rem;
          `}
          buttonId="join-button"
          width={20}
          height={3}
          fontSize={1.2}
        >
          참여하기
        </Button>
      </div>
    </>
  );
}
