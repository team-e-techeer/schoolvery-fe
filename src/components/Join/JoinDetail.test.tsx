import { render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import JoinDetail from './JoinDetail';
import React, { useState as useStateMock, useEffect as useEffectMock } from 'react';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(),
  useEffect: jest.fn(),
}));
describe('상세정보 페이지 테스트', () => {
  const setState = jest.fn();

  beforeEach(() => {
    (useStateMock as jest.Mock).mockImplementation(init => [init, setState]);
    (useEffectMock as jest.Mock).mockImplementation(f => f());
    jest.spyOn(React, 'useEffect');
    jest.spyOn(React, 'useState');

    jest.useFakeTimers().setSystemTime(1653480974069);
  });

  it('렌더링 테스트', () => {
    jest.useFakeTimers();
    jest.spyOn(global, 'setInterval');
    Date.now = jest.fn(() => 1653480974069);
    const { container } = render(
      <JoinDetail
        shopName="BBQ"
        schoolName="한국공과"
        categoryName="치킨"
        people={{ current: 3, total: 7 }}
        time={{ post: Date.now(), left: Date.now() }}
        payment={3000}
      />,
      { wrapper: BrowserRouter }
    );

    expect(screen.getByText('BBQ')).toBeInTheDocument();
    expect(screen.getByText('한국공과')).toBeInTheDocument();
    expect(screen.getByText('치킨')).toBeInTheDocument();
    expect(screen.getByText('현재 3명')).toBeInTheDocument();
    expect(screen.getByText('7명')).toBeInTheDocument();
    expect(screen.getByText('총 3000원')).toBeInTheDocument();
    expect(screen.getByText(`인당 ${Math.round(3000 / 7)}원 예상`)).toBeInTheDocument();
    expect(setInterval).toHaveBeenCalledTimes(1);
    expect(setInterval).toHaveBeenLastCalledWith(expect.any(Function), 1000 * 60);
    expect(useEffectMock).toHaveBeenCalledTimes(1);
    expect(useStateMock).toHaveBeenCalledTimes(3);
    expect(useEffectMock).toBeCalledTimes(1);

    expect(container).toMatchSnapshot();
  });
});
