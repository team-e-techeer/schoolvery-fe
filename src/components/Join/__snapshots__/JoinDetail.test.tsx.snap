// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`로그인 페이지 테스트 렌더링 테스트 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 90%;
  margin: 0 auto;
  background-color: #f2f4f6;
  border-radius: 8px;
  padding: 2rem;
  margin-top: 10rem;
  box-shadow: 2px 2px 2px 2px #d1d6db;
}

.emotion-1 {
  font-size: 2rem;
  margin-bottom: 1rem;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 1.3rem;
  margin-top: 1rem;
}

.emotion-3 {
  margin-right: 2rem;
  font-size: 2rem;
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 1.3rem;
  margin-top: 1rem;
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: 1.3rem;
}

.emotion-8:nth-of-type(2) {
  color: #3182f6;
}

.emotion-18 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  width: 20%;
  margin: 0 auto;
  height: 3rem;
  border: 0 solid transparent;
  border-radius: 0.7rem;
  background-color: #fc775a;
  color: #ffffff;
  font-size: 1.2rem;
  font-weight: 600;
  white-space: nowrap;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-font-smoothing: antialiased;
  -webkit-transition: color 0.1s ease-in-out,background-color 0.1s ease-in-out;
  transition: color 0.1s ease-in-out,background-color 0.1s ease-in-out;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-self: flex-end;
  -ms-flex-item-align: flex-end;
  align-self: flex-end;
  margin: 0;
  margin-top: 2rem;
}

.emotion-18:focus {
  outline: none;
}

.emotion-18:disabled {
  opacity: 0.26;
  cursor: not-allowed;
}

.emotion-18:active {
  background-color: #7b1fa2;
}

<div>
  <div
    class="emotion-0"
  >
    <h1
      class="emotion-1"
    >
      BBQ
    </h1>
    <span
      class="emotion-2"
    >
      <svg
        class="emotion-3"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M12 12c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2zm6-1.8C18 6.57 15.35 4 12 4s-6 2.57-6 6.2c0 2.34 1.95 5.44 6 9.14 4.05-3.7 6-6.8 6-9.14zM12 2c4.2 0 8 3.22 8 8.2 0 3.32-2.67 7.25-8 11.8-5.33-4.55-8-8.48-8-11.8C4 5.22 7.8 2 12 2z"
        />
      </svg>
      한국공과
    </span>
    <span
      class="emotion-2"
    >
      <svg
        class="emotion-3"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 3H4a1 1 0 0 0-1 1v6a1 1 0 0 0 1 1h6a1 1 0 0 0 1-1V4a1 1 0 0 0-1-1zM9 9H5V5h4v4zm11-6h-6a1 1 0 0 0-1 1v6a1 1 0 0 0 1 1h6a1 1 0 0 0 1-1V4a1 1 0 0 0-1-1zm-1 6h-4V5h4v4zm-9 4H4a1 1 0 0 0-1 1v6a1 1 0 0 0 1 1h6a1 1 0 0 0 1-1v-6a1 1 0 0 0-1-1zm-1 6H5v-4h4v4zm8-6c-2.206 0-4 1.794-4 4s1.794 4 4 4 4-1.794 4-4-1.794-4-4-4zm0 6c-1.103 0-2-.897-2-2s.897-2 2-2 2 .897 2 2-.897 2-2 2z"
        />
      </svg>
      치킨
    </span>
    <ul
      class="emotion-6"
    >
      <svg
        class="emotion-3"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M258.9 48C141.92 46.42 46.42 141.92 48 258.9c1.56 112.19 92.91 203.54 205.1 205.1 117 1.6 212.48-93.9 210.88-210.88C462.44 140.91 371.09 49.56 258.9 48zm-3.68 152.11c.21-1.2.44-2.4.71-3.59a66.46 66.46 0 0116.29-31.21c12.89-13.73 31.16-21.31 51.45-21.31a74.05 74.05 0 0125.06 4.26 66.69 66.69 0 0126.27 17.2 68.15 68.15 0 0118 42.14 78.46 78.46 0 010 11.4 86.19 86.19 0 01-8.2 31q-.76 1.59-1.59 3.15c-1.11 2.07-2.3 4.1-3.58 6.06a79.47 79.47 0 01-8.63 11c-13.12 14-29.92 21.73-47.31 21.73a59.61 59.61 0 01-19.17-3.18 63.47 63.47 0 01-6.1-2.43 70.76 70.76 0 01-22.07-16.12 83.76 83.76 0 01-22-51.32q-.27-3.88-.18-7.68a75.62 75.62 0 011.05-11.08zm-149.73 24.34a59.87 59.87 0 015.2-20.64 56.76 56.76 0 012.78-5.3 54.49 54.49 0 017.19-9.56 55.62 55.62 0 0114-10.82 56.84 56.84 0 018.11-3.64 63.85 63.85 0 0133.35-2.39 57 57 0 0130.78 17 57.86 57.86 0 0115.41 38.62c.05 2.11 0 4.23-.15 6.38a71.58 71.58 0 01-6 23.84 69.49 69.49 0 01-5.73 10.42 65.39 65.39 0 01-15.76 16.57c-1.5 1.07-3.06 2.07-4.67 3.07a54.21 54.21 0 01-10 4.65 49.31 49.31 0 01-16.2 2.76c-.93 0-1.86 0-2.78-.08a47.6 47.6 0 01-5.48-.62 51.19 51.19 0 01-5.35-1.23 53.54 53.54 0 01-7.72-2.89c-.84-.39-1.66-.8-2.48-1.23-18-9.49-31.57-29.16-34.23-52.12-.12-1.05-.22-2.1-.29-3.16a66.59 66.59 0 01.02-9.63zm53.92 178.6a177.27 177.27 0 01-61.94-70.65 4 4 0 011.62-5.26C117.67 316.69 141.4 311 163.82 311c17 0 30.7 2 42.69 5.88a8 8 0 012.59 13.77c-23.35 19-38.4 42.54-45.47 70.75a2.77 2.77 0 01-4.22 1.65zM256 432a175.12 175.12 0 01-65.7-12.72 4 4 0 01-2.4-4.46c.4-2.05.84-3.92 1.23-5.48 7.12-28.43 24.76-52 51-68.18 23.29-14.35 53-22.25 83.52-22.25 31.16 0 60 7.58 83.48 21.91a2.72 2.72 0 01.91 3.67A176.1 176.1 0 01256 432z"
        />
        <path
          d="M161 295.28a47.6 47.6 0 01-5.48-.62 47.6 47.6 0 005.48.62zm-26.36-117.15a55.62 55.62 0 00-14 10.82 54.49 54.49 0 00-7.19 9.56 54.49 54.49 0 017.19-9.56 55.62 55.62 0 0114-10.82zm81.53 79.76a71.58 71.58 0 006-23.84c.15-2.15.2-4.27.15-6.38q.08 3.15-.15 6.38a71.58 71.58 0 01-6 23.84zm-81.53-79.76a56.84 56.84 0 018.11-3.64 56.84 56.84 0 00-8.11 3.64zm15.57 115.3a53.54 53.54 0 01-7.72-2.89 53.54 53.54 0 007.72 2.89zm-44.43-56.24c2.66 23 16.26 42.63 34.23 52.12-18.01-9.49-31.57-29.16-34.23-52.12zM254.34 219a83.76 83.76 0 0022 51.32 70.76 70.76 0 0022.07 16.12 70.76 70.76 0 01-22.07-16.12 83.76 83.76 0 01-22-51.32q-.27-3.88-.18-7.68-.09 3.75.18 7.68zm50.16 69.82a63.47 63.47 0 01-6.1-2.43 63.47 63.47 0 006.1 2.43zm-48.57-92.28a66.46 66.46 0 0116.29-31.21 66.46 66.46 0 00-16.29 31.21zM375 165.46a68.15 68.15 0 0118 42.14 68.15 68.15 0 00-18-42.14 66.69 66.69 0 00-26.27-17.2 66.69 66.69 0 0126.27 17.2zM393 219a86.19 86.19 0 01-8.2 31 86.19 86.19 0 008.2-31zm-138.84-7.73a75.62 75.62 0 011.06-11.14 75.62 75.62 0 00-1.06 11.14zm129.03 41.89zm-176.31-64.11a57.86 57.86 0 0115.41 38.62 57.86 57.86 0 00-15.41-38.62 57 57 0 00-30.78-17 57 57 0 0130.78 17zM190 288a54.21 54.21 0 01-10 4.65 54.21 54.21 0 0010-4.65zm-84.51-63.55a59.87 59.87 0 015.2-20.64 59.87 59.87 0 00-5.2 20.64zm89.19 60.43C193.17 286 191.61 287 190 288c1.61-1 3.17-2 4.68-3.12zm21.49-26.99a69.49 69.49 0 01-5.73 10.42 69.49 69.49 0 005.73-10.42zm-105.48-54.08a56.76 56.76 0 012.78-5.3 56.76 56.76 0 00-2.78 5.3zm83.99 81.07a65.39 65.39 0 0015.76-16.57 65.39 65.39 0 01-15.76 16.57z"
        />
      </svg>
      <span
        class="emotion-8"
      >
        7
        명
      </span>
      <span
        class="emotion-8"
      >
        현재 
        3
        명
      </span>
    </ul>
    <ul
      class="emotion-6"
    >
      <svg
        class="emotion-3"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 64C150 64 64 150 64 256s86 192 192 192 192-86 192-192S362 64 256 64z"
          fill="none"
          stroke-miterlimit="10"
          stroke-width="32"
        />
        <path
          d="M256 128v144h96"
          fill="none"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="32"
        />
      </svg>
      <span
        class="emotion-8"
      >
        오후 09 : 16
      </span>
      <span
        class="emotion-8"
      >
        오후 09 : 16
         남음
      </span>
    </ul>
    <ul
      class="emotion-6"
    >
      <svg
        class="emotion-3"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 640 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 448c0 17.67 14.33 32 32 32h576c17.67 0 32-14.33 32-32V128H0v320zm448-208c0-8.84 7.16-16 16-16h96c8.84 0 16 7.16 16 16v32c0 8.84-7.16 16-16 16h-96c-8.84 0-16-7.16-16-16v-32zm0 120c0-4.42 3.58-8 8-8h112c4.42 0 8 3.58 8 8v16c0 4.42-3.58 8-8 8H456c-4.42 0-8-3.58-8-8v-16zM64 264c0-4.42 3.58-8 8-8h304c4.42 0 8 3.58 8 8v16c0 4.42-3.58 8-8 8H72c-4.42 0-8-3.58-8-8v-16zm0 96c0-4.42 3.58-8 8-8h176c4.42 0 8 3.58 8 8v16c0 4.42-3.58 8-8 8H72c-4.42 0-8-3.58-8-8v-16zM624 32H16C7.16 32 0 39.16 0 48v48h640V48c0-8.84-7.16-16-16-16z"
        />
      </svg>
      <span
        class="emotion-8"
      >
        총 
        3000
        원
      </span>
      <span
        class="emotion-8"
      >
        인당 
        429
        원 예상
      </span>
    </ul>
    <button
      class="emotion-18"
      id="join-button"
    >
      <span>
        참여하기
      </span>
    </button>
  </div>
</div>
`;
